File: options_lab/strategy/expiry_put.py
"""Expiry-day short put — the one structure in this project that measured
positive on both train and a sealed holdout.

    On a NIFTY expiry day, at 11:00 IST, sell one put at the nearest listed
    strike to F x (1 - otm_pct), where F is the put-call-parity forward.
    Hold to cash settlement. Never buy it back.

Measured on 170 expiry sessions 2023-01-05..2026-04-13, lot 65, all costs:
0.75% OTM gives 97.65% win rate and +Rs 406/trade; 1.00% OTM gives 100.00%
and +Rs 270. The 0.75% variant pays more AND has actually shown its losses,
which is why it is the default here.

WHY THIS IS NOT PAYOFF GEOMETRY: every configuration that reached 100% is a
naked short PUT. Calls reach 100% at no distance out to 3.5% and no strangle
reaches it. The record is the statement "NIFTY did not fall 1% between 11:00
and the close on 170 consecutive expiries" - a property of a 2023-2026 bull
market, not of the payoff. The loss is unbounded; size accordingly.
"""
from __future__ import annotations

from datetime import date, time

import numpy as np
import pandas as pd

from options_lab import costs

DEFAULT_OTM_PCT = 0.0075        # 0.75%: pays more than the 100% cell at 1.00%
DEFAULT_N_STRIKES = 10          # pinned, so the forward is deterministic
SETTLE_FROM = time(15, 0)       # NSE settles on the last-30-minute average
SETTLE_TO = time(15, 30)


class ThinChain(ValueError):
    """Too few strikes quote both sides to compute a trustworthy forward."""


class NotASnapshot(ValueError):
    """The chain carries a strike more than once.

    The raw chain has one row per contract PER MINUTE, so passing it whole is
    the natural mistake. Before this check it surfaced as pandas' "The truth
    value of a Series is ambiguous" from deep inside the comparison, which
    tells the caller nothing. Take one timestamp's snapshot first, e.g.
    `chain[chain.ts <= entry].sort_values("ts").groupby(["strike","right"],
    as_index=False).last()`.
    """


def parity_forward(chain: pd.DataFrame, *, n_strikes: int = DEFAULT_N_STRIKES) -> float:
    """F = median over the n nearest strikes of (K + CE - PE).

    Read off the OPTIONS, never the future. The near-month future carries a
    median +48.5 point basis over spot (max +228), which is most of a 0.75%
    strike offset - using it would mis-set the strike every single day.
    """
    dupes = chain.duplicated(subset=["strike", "right"], keep=False)
    if dupes.any():
        offenders = sorted(chain.loc[dupes, "strike"].unique())[:3]
        raise NotASnapshot(
            f"chain is not a snapshot: strikes {offenders} appear more than "
            f"once ({int(dupes.sum())} duplicated rows). Reduce to one "
            f"timestamp per contract before computing the forward."
        )

    ce = chain[chain["right"] == "CE"].set_index("strike")["close"]
    pe = chain[chain["right"] == "PE"].set_index("strike")["close"]
    both = ce.index.intersection(pe.index)
    usable = [k for k in both if ce[k] > 0 and pe[k] > 0]

    if len(usable) < n_strikes:
        raise ThinChain(
            f"{len(usable)} strikes quote both sides; need {n_strikes}"
        )

    implied = pd.Series({k: k + float(ce[k]) - float(pe[k]) for k in usable})
    # Two passes: a rough centre from all strikes, then the n nearest to it.
    centre = float(implied.median())
    nearest = implied.reindex(
        sorted(implied.index, key=lambda k: abs(k - centre))[:n_strikes]
    )
    return float(nearest.median())


def select_strike(strikes, *, forward: float,
                  otm_pct: float = DEFAULT_OTM_PCT) -> float:
    """The NEAREST listed strike to the target - not the nearest at or below it.

    That distinction is worth Rs 81/trade, roughly a fifth of the strategy's
    return: flooring collects a smaller credit (Rs 239 vs Rs 320 median) for a
    marginally higher win rate. Ties break to the lower (further OTM) strike.
    """
    ladder = sorted(float(k) for k in strikes)
    if not ladder:
        raise ValueError("empty strike ladder")
    target = forward * (1.0 - otm_pct)
    return min(ladder, key=lambda k: (abs(k - target), k))


def settlement_price(spot: pd.Series, day: date) -> float:
    """NSE settles NIFTY options on the weighted average of spot over the last
    thirty minutes, NOT the 15:29 print. Using the print flips which sessions
    lose: 2026-01-20 drops out and 2024-04-18 drops in."""
    idx = pd.DatetimeIndex(spot.index)
    window = spot[(idx.date == day) & (idx.time >= SETTLE_FROM) & (idx.time < SETTLE_TO)]
    if window.empty:
        raise ValueError(f"no bars in the settlement window on {day}")
    return float(window.mean())


def settle_trade(*, strike: float, credit: float, settlement: float,
                 lot_size: int, lots: int = 1, regime: str = "quoted") -> dict:
    """P&L of one short put carried to cash settlement.

    The short leg pays no exercise STT - that 0.125%-of-intrinsic charge falls
    on the buyer. The loss is unbounded below and scales one-for-one with the
    move past the strike.
    """
    intrinsic = max(strike - settlement, 0.0)
    qty = lot_size * lots
    gross = (credit - intrinsic) * qty
    charges = costs.sell_to_settle(premium=credit, lot_size=lot_size, lots=lots,
                                   regime=regime)
    net = gross - charges.total
    return {
        "strike": strike, "credit": credit, "settlement": settlement,
        "intrinsic": intrinsic, "qty": qty,
        "gross_pnl": gross, "cost": charges.total, "net_pnl": net,
        "won": net > 0,
    }

File: android/engine/src/main/kotlin/com/optionslab/engine/ExpiryPut.kt
package com.optionslab.engine

import java.time.LocalDate
import kotlin.math.abs
import kotlin.math.max

/**
 * Expiry-day short put. Port of `strategy/expiry_put.py`.
 *
 *     On a NIFTY expiry day, at 11:00 IST, sell one put at the nearest listed
 *     strike to F x (1 - otm_pct), where F is the put-call-parity forward.
 *     Hold to cash settlement. Never buy it back.
 *
 * The record is the statement "NIFTY did not fall 1% between 11:00 and the
 * close on 170 consecutive expiries" - a property of a 2023-2026 bull market,
 * not of the payoff. The loss is unbounded; size accordingly.
 */
object ExpiryPut {
    const val DEFAULT_OTM_PCT = 0.0075
    const val DEFAULT_N_STRIKES = 10
    const val SETTLE_FROM = 15 * 60          // NSE settles on the last-30-minute average
    const val SETTLE_TO = 15 * 60 + 30
    const val DEFAULT_ENTRY = 11 * 60
    const val DEFAULT_HOLDOUT_SESSIONS = 30

    class ThinChain(msg: String) : IllegalArgumentException(msg)
    class NotASnapshot(msg: String) : IllegalArgumentException(msg)
    class SessionSkipped(msg: String) : Exception(msg)

    /** One contract's price at one moment - a row of a snapshot. */
    data class Quote(val strike: Double, val right: Right, val close: Double)

    /** F = median over the n nearest strikes of (K + CE - PE). Read off the OPTIONS, never the future. */
    fun parityForward(snapshot: List<Quote>, nStrikes: Int = DEFAULT_N_STRIKES): Double {
        val seen = HashMap<Pair<Double, Right>, Int>()
        for (q in snapshot) seen.merge(q.strike to q.right, 1, Int::plus)
        val dupes = seen.filter { it.value > 1 }
        if (dupes.isNotEmpty()) {
            val offenders = dupes.keys.map { it.first }.distinct().sorted().take(3)
            throw NotASnapshot(
                "chain is not a snapshot: strikes $offenders appear more than once (${dupes.values.sum()} duplicated rows). " +
                    "Reduce to one timestamp per contract before computing the forward.")
        }
        val ce = LinkedHashMap<Double, Double>()
        val pe = LinkedHashMap<Double, Double>()
        for (q in snapshot) when (q.right) {
            Right.CE -> ce[q.strike] = q.close
            Right.PE -> pe[q.strike] = q.close
            Right.IX -> {}
        }
        val usable = ce.keys.filter { it in pe }.sorted().filter { ce.getValue(it) > 0 && pe.getValue(it) > 0 }
        if (usable.size < nStrikes) throw ThinChain("${usable.size} strikes quote both sides; need $nStrikes")
        val implied = usable.map { k -> k to (k + ce.getValue(k) - pe.getValue(k)) }
        val centre = Stats.median(implied.map { it.second })
        val nearest = implied.sortedBy { abs(it.first - centre) }.take(nStrikes)
        return Stats.median(nearest.map { it.second })
    }

    /** The NEAREST listed strike to the target; ties break to the lower strike. */
    fun selectStrike(strikes: Collection<Double>, forward: Double, otmPct: Double = DEFAULT_OTM_PCT): Double {
        require(strikes.isNotEmpty()) { "empty strike ladder" }
        val target = forward * (1.0 - otmPct)
        return strikes.sorted().minWith(compareBy<Double>({ abs(it - target) }, { it }))
    }

    /** NSE settles on the average of spot over 15:00-15:29, not the 15:29 print. */
    fun settlementPrice(spot: Map<Int, Double>): Double {
        val window = spot.filterKeys { it in SETTLE_FROM until SETTLE_TO }.values
        if (window.isEmpty()) throw IllegalArgumentException("no bars in the settlement window")
        return window.average()
    }

    data class Trade(
        val session: LocalDate?,
        val strike: Double,
        val credit: Double,          // net credit per unit
        val settlement: Double,
        val intrinsic: Double,
        val lotSize: Int,
        val lots: Int,
        val qty: Int,
        val wingStrike: Double?,
        val wingDebit: Double,
        val wingIntrinsic: Double,
        val maxLoss: Double?,
        val grossPnl: Double,
        val cost: Double,
        val netPnl: Double,
        val forward: Double = Double.NaN,
        val otmRealised: Double = Double.NaN,
    ) {
        val won: Boolean get() = netPnl > 0
    }

    /** P&L of one short put - naked, or spread against a bought wing. */
    fun settleTrade(
        strike: Double, credit: Double, settlement: Double, lotSize: Int, lots: Int = 1,
        regime: String = "quoted", wingStrike: Double? = null, wingDebit: Double = 0.0,
    ): Trade {
        val qty = lotSize * lots
        val shortIntrinsic = max(strike - settlement, 0.0)
        var cost = Costs.sellToSettle(credit, lotSize, lots, regime).total
        var wingIntrinsic = 0.0
        if (wingStrike != null) {
            wingIntrinsic = max(wingStrike - settlement, 0.0)
            cost += Costs.buyToSettle(wingDebit, lotSize, lots, regime, wingIntrinsic).total
        }
        val netCredit = credit - wingDebit
        val intrinsic = shortIntrinsic - wingIntrinsic
        val gross = (netCredit - intrinsic) * qty
        val net = gross - cost
        return Trade(
            session = null, strike = strike, credit = netCredit, settlement = settlement,
            intrinsic = intrinsic, lotSize = lotSize, lots = lots, qty = qty,
            wingStrike = wingStrike, wingDebit = wingDebit, wingIntrinsic = wingIntrinsic,
            maxLoss = if (wingStrike == null) null else (strike - wingStrike - netCredit) * qty + cost,
            grossPnl = gross, cost = cost, netPnl = net,
        )
    }

    /** The latest bar per (strike, right) at or before [minute]. */
    fun snapshot(chain: List<Series>, minute: Int): List<Quote> = chain.mapNotNull { s ->
        if (s.right == Right.IX) return@mapNotNull null
        val i = s.lastAtOrBefore(minute)
        if (i < 0) null else Quote(s.strike, s.right, s.close[i])
    }.sortedWith(compareBy({ it.strike }, { it.right.ordinal }))

    fun minuteSlice(chain: List<Series>, minute: Int): List<Quote> = chain.mapNotNull { s ->
        if (s.right == Right.IX) return@mapNotNull null
        val i = s.indexOf(minute)
        if (i < 0) null else Quote(s.strike, s.right, s.close[i])
    }

    /** "dated": each session's own lot - the chain first, the NSE table as fallback. */
    fun datedLot(session: Session, underlying: String = "NIFTY"): Int =
        session.lotHint ?: Lots.lotFromChain(session.series) ?: Lots.lotSizeOn(underlying, session.day)

    sealed interface LotChoice {
        data class Pinned(val lot: Int) : LotChoice
        data object Dated : LotChoice
    }

    data class Params(
        val otmPct: Double = DEFAULT_OTM_PCT,
        val entryMinute: Int = DEFAULT_ENTRY,
        val lots: Int = 1,
        val regime: String = "quoted",
        val wingPct: Double? = null,
        val lot: LotChoice = LotChoice.Pinned(65),
        val underlying: String = "NIFTY",
    )

    /** One expiry session -> one trade. Throws SessionSkipped with a reason. */
    fun runSession(day: LocalDate, chain: List<Series>, lotSize: Int, p: Params): Trade {
        val opts = chain.filter { it.right != Right.IX }
        if (opts.none { s -> s.size > 0 && s.minutes[0] <= p.entryMinute }) {
            throw SessionSkipped("no bars at or before ${minuteText(p.entryMinute)}")
        }
        val snap = snapshot(opts, p.entryMinute)
        val forward = parityForward(snap)
        val strikes = snap.map { it.strike }.distinct()
        val strike = selectStrike(strikes, forward, p.otmPct)
        val leg = snap.firstOrNull { it.strike == strike && it.right == Right.PE }
        if (leg == null || leg.close <= 0) throw SessionSkipped("no PE quote at strike ${fmtG(strike)}")
        val credit = leg.close

        val windowMinutes = opts.flatMap { s -> s.minutes.filter { it in SETTLE_FROM until SETTLE_TO } }.toSortedSet()
        if (windowMinutes.isEmpty()) throw SessionSkipped("no bars in the settlement window")
        val forwards = ArrayList<Double>()
        for (m in windowMinutes) {
            try { forwards += parityForward(minuteSlice(opts, m)) } catch (_: ThinChain) {} catch (_: NotASnapshot) {}
        }
        if (forwards.isEmpty()) throw SessionSkipped("settlement window too thin to price")

        var wingStrike: Double? = null
        var wingDebit: Double? = null
        if (p.wingPct != null) {
            require(p.wingPct >= 0) {
                "wing_pct must be positive; ${p.wingPct} would buy a put ABOVE the short strike, which is not a hedge"
            }
            val ws = selectStrike(strikes, forward, p.otmPct + p.wingPct)
            if (ws >= strike) throw SessionSkipped(
                "wing at ${fmtG(ws)} is not below the short strike ${fmtG(strike)}; a zero-width spread is two orders and no hedge")
            val wl = snap.firstOrNull { it.strike == ws && it.right == Right.PE }
            if (wl == null || wl.close <= 0) throw SessionSkipped("no PE quote at wing strike ${fmtG(ws)}")
            wingStrike = ws
            wingDebit = wl.close
        }
        val t = settleTrade(strike, credit, forwards.average(), lotSize, p.lots, p.regime, wingStrike, wingDebit ?: 0.0)
        return t.copy(session = day, forward = forward, otmRealised = (forward - strike) / forward)
    }

    data class Skip(val day: LocalDate, val why: String)

    /** Run every session. Skips are never silent. */
    fun runBacktest(sessions: List<Session>, p: Params): Pair<List<Trade>, List<Skip>> {
        val trades = ArrayList<Trade>()
        val skipped = ArrayList<Skip>()
        for (s in sessions) {
            try {
                val lot = when (val l = p.lot) {
                    is LotChoice.Pinned -> l.lot
                    LotChoice.Dated -> datedLot(s, p.underlying)
                }
                trades += runSession(s.day, s.series, lot, p)
            } catch (e: SessionSkipped) {
                skipped += Skip(s.day, (e.message ?: "").take(80))
            } catch (e: ThinChain) {
                skipped += Skip(s.day, (e.message ?: "").take(80))
            } catch (e: NotASnapshot) {
                skipped += Skip(s.day, (e.message ?: "").take(80))
            } catch (e: Lots.NoLotSize) {
                skipped += Skip(s.day, (e.message ?: "").take(80))
            }
        }
        return trades to skipped
    }

    /** Seal the most recent `holdout` sessions. Chronological, never random. */
    fun splitSessions(days: List<LocalDate>, holdout: Int = DEFAULT_HOLDOUT_SESSIONS): Pair<List<LocalDate>, List<LocalDate>> {
        val ordered = days.toSortedSet().toList()
        require(holdout < ordered.size) {
            "holdout of $holdout leaves no training sessions out of ${ordered.size}; a split needs data on both sides"
        }
        return ordered.dropLast(holdout) to ordered.takeLast(holdout)
    }
}
